### PR TITLE
[refactor] interfaces: Add missing LockAnnotation for cs_main

### PR DESCRIPTION
**Pull request by maflcko**
This adds missing `LockAnnotation lock(::cs_main);` to `src/interfaces/chain.cpp` (as well as tests and  benchmarks)

### CODE REVIEW

File: src/wallet/test/wallet_tests.cpp [hunk 1]
```diff
@@ -245,9 +246,11 @@ BOOST_FIXTURE_TEST_CASE(importwallet_rescan, TestChain100Setup)
 BOOST_FIXTURE_TEST_CASE(coin_mark_dirty_immature_credit, TestChain100Setup)
 {
     auto chain = interfaces::MakeChain();
+    LockAnnotation lock(::cs_main);
+    auto locked_chain = chain->lock();
```
Comment (promag) [inline src/wallet/test/wallet_tests.cpp]: Why was this moved?

File: src/test/txvalidationcache_tests.cpp [hunk 2]
```diff
@@ -66,29 +66,41 @@ BOOST_FIXTURE_TEST_CASE(tx_mempool_block_doublespend, TestChain100Setup)
...
 
     // Test 2: ... and should be rejected if spend1 is in the memory pool
     BOOST_CHECK(ToMemPool(spends[0]));
     block = CreateAndProcessBlock(spends, scriptPubKey);
-    BOOST_CHECK(chainActive.Tip()->GetBlockHash() != block.GetHash());
-    mempool.clear();
+    {
+        LOCK(cs_main);
+        BOOST_CHECK(chainActive.Tip()->GetBlockHash() != block.GetHash());
+        mempool.clear();
```
Comment (promag) [inline src/test/txvalidationcache_tests.cpp]: This gives the bad impression `cs_main` is necessary.

**Review by promag — COMMENTED** [review]
Concept ACK.

**Comment by MarcoFalke** [comment]
Addressed feedback by @promag

**Comment by practicalswift** [comment]
Concept ACK

**Comment by promag** [comment]
utACK face05a080c081fc426ff34ecb39ac072710034b.

File: src/wallet/test/wallet_tests.cpp [hunk 3]
```diff
@@ -37,14 +37,14 @@ static void AddKey(CWallet& wallet, const CKey& key)
 BOOST_FIXTURE_TEST_CASE(scan_for_wallet_transactions, TestChain100Setup)
 {
     auto chain = interfaces::MakeChain();
+    LockAnnotation lock(::cs_main);
```
Comment (practicalswift) [inline src/wallet/test/wallet_tests.cpp]: I might be missing something but shouldn't `LockAnnotation` be used only when we are holding the mutex but the compiler thread analysis doesn't understand it? In this case we aren't holding `cs_main` here, right?

If it is used simply to opt-out of thread-safety analysis (when we are not holding the mutex) then `NO_THREAD_SAFETY_ANALYSIS` should be used, no?

Comment (MarcoFalke) [reply src/wallet/test/wallet_tests.cpp]: The tests are single threaded, so taking a lock is equivalent to pretending it is taken.

Not sure how I can use `NO_THREAD_SAFETY_ANALYSIS` to achieve that.

Comment (practicalswift) [reply src/wallet/test/wallet_tests.cpp]: Yes, I know the tests are single-threaded but I think it might be confusing for newcomers if we use `LockAnnotation` to both a.) guarantee to the compiler that the lock is taken (like in `src/interfaces/chain.cpp`), and b.) to opt-out of thread safety analysis when the lock is not taken (like in this file).

Just to make it crystal clear to readers what we are doing we could use different names for these two use-cases? (If using `NO_THREAD_SAFETY_ANALYSIS` is not an option.)

Something like `struct SCOPED_LOCKABLE DisableThreadSafetyAnalysis` (like `LockAnnotation`) or `#define DISABLE_THREAD_SAFETY_ANALYSIS(cs, code) [&] NO_THREAD_SAFETY_ANALYSIS { code; }()`?

Comment (MarcoFalke) [reply src/wallet/test/wallet_tests.cpp]: Added a comment instead. Is that fine?

**Comment by MarcoFalke** [comment]
> Perhaps --enable-debug builds should do AssertLockHeld(…) as part of the LockAnnotationctor to make sure the LockAnnotation guarantees hold up over time

Sounds good. I have addressed your concerns, but this suggestion should go into a follow-up pull request?

**Review by ryanofsky — APPROVED** [review]
utACK fa3c6511435149782545ac0d09d4722dc115d709

**Comment by practicalswift** [comment]
utACK fa3c6511435149782545ac0d09d4722dc115d709
